### PR TITLE
Stripe Integration

### PR DESCRIPTION
**Pull request by ericpias**
This is V1 of stripe integration and supports payment method(s) setup only.
See functional requirements: https://docs.google.com/document/d/1etw2-A-zjceqvqJmNJYBKNSX6unxCf3Gnd5jVbyqS8I/edit?usp=sharing
See Tech Integration: https://docs.google.com/document/d/1vqrKkI-iF4se0JcBQjxyLl8vb_-yxpC3lIqIe0we3-k/edit?usp=sharing

Test Outline:

- Optional PM field and no card entered - does NOT setup
- Optional PM field and card entered - does setup
- Required PM field and no card entered - required field error
- Required PM field and card entered - does setup
- Card entry incomplete - cannot submit - error shown
- Test live form and review and make sure the proper stripe key is used.  Simulate using two stripe accounts.  Use hosted forms with diff feathery api keys.  Using 2 stripe accounts ([eric.pias@comcast.net](mailto:eric.pias@comcast.net) (PW saved in Safari) and [ericdpias@gmail.com](mailto:ericdpias@gmail.com) (FF)).
- Multi-step form
- Setup card and then go back, must setup PM again if required or card entered.
- Verify card not setup on steps where card does not appear.
- Test multiple cards  (must both go to same customer)
- Two cards one step
- Two steps, each with a PM field
- Two steps, first 2 cards, 2nd 1 card.
- Test a form with no PM field
- Test DB master against stripe Lib and BE.

### CODE REVIEW

File: src/elements/fields/PaymentMethodField.jsx [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  'lineHeight',
+  'letterSpacing',
+  'textAlign',
+  'padding',
+  'textDecoration',
+  'textShadow',
+  'textTransform'
+];
+
+//
```
Comment (obmelvin) [inline src/elements/fields/PaymentMethodField.jsx:37]: empty comment :)

File: src/elements/fields/PaymentMethodField.jsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  useStripe
+} from '@stripe/react-stripe-js';
+
+import { getStripePromise } from '../../integrations/stripe';
+import { filterStyleValues } from '../../utils/styles';
+import { filterKeys } from '../../utils/utils';
+
+import InlineTooltip from '../components/Tooltip';
+
+const stripeKey = 'any-key-does-not-matter-what';
```
Comment (obmelvin) [inline src/elements/fields/PaymentMethodField.jsx:15]: not sure if this was needed during dev, or still needed. would benefit from a comment explaining why you need a dummy key

File: src/elements/fields/PaymentMethodField.jsx [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+      }
+    },
+    classes: {
+      base: element.servar.key
+    }
+  };
+  const handleCardChange = async (event) => {
+    // Listen for changes in the CardElement
+    // and display any errors as the customer types their card details
+    setFieldError(event.error ? event.error.message : '');
```
Comment (obmelvin) [inline src/elements/fields/PaymentMethodField.jsx:98]: you could do `event.error?.message ?? ''`

File: src/elements/fields/index.jsx [hunk 4]
```diff
@@ -189,6 +191,37 @@ function applyFieldStyles(field, styles) {
       styles.applyCorners('field');
       styles.applyBorders('field');
       break;
+    case 'payment_method':
+      styles.addTargets('sub-fc', 'tooltipIcon', 'completed');
+      styles.applyHeight('sub-fc');
+      // styles.applyWidth('sub-fc');
```
Comment (obmelvin) [inline src/elements/fields/index.jsx:197]: double check if you need this style or not

File: src/integrations/stripe.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
+import { loadStripe } from '@stripe/stripe-js';
```
Comment (obmelvin) [inline src/integrations/stripe.js:1]: the other integrations are all loaded async rather than being installed in package.json. can we not do that here?

File: webpack.development.js [hunk 6]
```diff
@@ -1,19 +1,19 @@
...
-module.exports = config;
+  config.devtool = 'eval-cheap-module-source-map';
+  config.externals = ['react'];
+  config.plugins.push(
+    new webpack.optimize.LimitChunkCountPlugin({
+      maxChunks: 1
+    })
+  );
+  return config;
+};
```
Comment (obmelvin) [inline webpack.development.js:19]: not too important, but apparently missing new line at end of file

File: src/elements/fields/PaymentMethodField.jsx (hunk 2 above)
Comment (ericpias) [reply src/elements/fields/PaymentMethodField.jsx:15]: Yes it is still used in the dashboard only.  Adding a comment.

File: src/elements/fields/index.jsx (hunk 4 above)
Comment (ericpias) [reply src/elements/fields/index.jsx:197]: Not needed, width is in 'fc' only.

File: src/integrations/stripe.js (hunk 5 above)
Comment (ericpias) [reply src/integrations/stripe.js:1]: Stripe-js is a very light weight wrapper lib that does really one thing and that is load stripe.js from a secure, PSI-compliant location.  Yes, we could duplicate that and I looked into doing that, but it is not trivial and why re-invent the wheel here so this is the way I went.  The uncompressed size of stripe-js is 3.4k.  Quite small.

Comment (ericpias) [reply src/integrations/stripe.js:1]: BTW, the actual loading of stripe.js (the larger and secure one) is done async and only if the integration is active.

Comment (obmelvin) [reply src/integrations/stripe.js:1]: ok cool, I saw that it was importing loadStripe but wasn't 100% sure if that was happening. I think I'm ok with this if @bo-dun-1 is

File: package.json [hunk 7]
```diff
@@ -15,6 +15,7 @@
     "build:node": "webpack --mode production --config webpack.node.js",
     "build:umd": "webpack --mode production --config webpack.umd.js",
     "dev": "webpack --mode development --config webpack.development.js --watch",
+    "bundle-analysis": "webpack --mode development --config webpack.development.js --env analyze",
```
Comment (bo-dun-1) [inline package.json:18]: analyze-bundle?

File: src/elements/fields/PaymentMethodField.jsx [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+      return;
+    }
+    setCardElement(element.servar.key, elements.getElement(CardElement));
+  }, [stripe, element.servar.key, elements]);
+
+  const cardElementOptions = {
+    style: {
+      base: {
+        textDecoration: 'none', // Bug in card element - force the reset to none
+        ...filterKeys(applyStyles.getTarget('field'), supportedStripeCssProps),
```
Comment (bo-dun-1) [inline src/elements/fields/PaymentMethodField.jsx:68]: does the stripe card element support mobile styles? in this current usage of applystyles we are not fetching the mobile styles, which are stored as an object in a media query

File: src/elements/fields/PaymentMethodField.jsx [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        '::placeholder': {
+          textDecoration: 'none', // Bug in card element - force the reset to none
+          ...filterKeys(
+            applyStyles.getTarget('placeholder'),
+            supportedStripeCssProps
+          )
+        },
+        ':hover': filterStyleValues(
```
Comment (bo-dun-1) [inline src/elements/fields/PaymentMethodField.jsx:76]: instead of retroactively filtering out teh !important suffix, it might be cleaner to pass in a flag to omit the !important flag when generating the hover/selected styles for applyStyles

File: src/elements/fields/PaymentMethodField.jsx [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+          filterKeys(applyStyles.getTarget('hover'), supportedStripeCssProps),
+          '!important'
+        ),
+        ':focus': filterStyleValues(
+          filterKeys(applyStyles.getTarget('active'), supportedStripeCssProps),
+          '!important'
+        )
+      },
+      complete: {
+        ...applyStyles.getTarget('completed')
```
Comment (bo-dun-1) [inline src/elements/fields/PaymentMethodField.jsx:86]: no need to be spread

File: src/elements/fields/PaymentMethodField.jsx [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+      },
+      invalid: {
+        color: '#d6504a',
```
Comment (bo-dun-1) [inline src/elements/fields/PaymentMethodField.jsx:89]: the error color we're using in the react lib right now is #F42525, e.g. with inline errors. we might want to make that a const in utils and use it everywhere in teh lib, including here

File: src/elements/fields/PaymentMethodField.jsx [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+          ...applyStyles.getTarget('sub-fc'),
+          '&:focus': applyStyles.getTarget('active'),
+          '&:hover': applyStyles.getTarget('hover'),
+          ...(inlineError ? { borderColor: '#F42525' } : {})
+        }}
+      >
+        <div css={{ width: '100%', position: 'relative' }}>
+          {/* position an input field under the card element to support html5 errors */}
+          {runtime && errorDisplayMode === 'html5' && (
+            <input
```
Comment (bo-dun-1) [inline src/elements/fields/PaymentMethodField.jsx:139]: we do something similar with a hdiden input field for hte file _upload element. we might want to unify those components

File: src/elements/fields/PaymentMethodField.jsx [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+    if (!runtime)
+      document.dispatchEvent(
+        new CustomEvent('stripe_key_loaded', {
+          detail: { key: stripeKey }
+        })
+      );
+  }, [runtime]);
+
+  return (
+    <>
```
Comment (bo-dun-1) [inline src/elements/fields/PaymentMethodField.jsx:185]: fragment is unnecessary here

File: src/integrations/stripe.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    promise.then((stripe) => {
+      resolve(stripe);
+    });
+  });
+});
+
+export function getStripePromise() {
+  return stripePromise;
+}
+export async function getStripe() {
```
Comment (bo-dun-1) [inline src/integrations/stripe.js:15]: the async and await can be removed here, and you can just call await getStripe()

File: src/integrations/stripe.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+const stripePromise = new Promise((resolve) => {
+  document.addEventListener('stripe_key_loaded', (e) => {
+    const promise = loadStripe(e.detail.key);
+    promise.then((stripe) => {
+      resolve(stripe);
+    });
+  });
+});
+
+export function getStripePromise() {
```
Comment (bo-dun-1) [inline src/integrations/stripe.js:12]: const getStripePromise = () => stripePromise;

File: src/integrations/stripe.js (hunk 14 above)
Comment (bo-dun-1) [reply src/integrations/stripe.js:15]: export const getStripe = () => getStripePromise()

but this also seems redundant with the above function

File: src/integrations/stripe.js [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+export function getStripeCustomerFieldValues(stripeConfig, fieldValues) {
+  if (!stripeConfig?.metadata?.customer_field_mappings) return {};
+  const customer = getObjectMappingValues(
+    stripeConfig.metadata.customer_field_mappings,
+    fieldValues
+  );
+  // the shipping object in customer must have a name property (stripe requires this),
+  //  but we don't want to make the mapping required.  So default it here.
+  customer.shipping &&
```
Comment (bo-dun-1) [inline src/integrations/stripe.js:38]: i would make this more explicit.

if (!customer.shipping?.name) customer.shipping.name = 'Name not supplied'

File: src/form/Form.tsx [hunk 17]
```diff
@@ -953,6 +962,63 @@ function Form({
     });
     if (invalid) return;
 
+    // Perform any Stripe payment method setup here if and only if the
+    // payment method field has a complete value
+    const paymentMethodFields = getPaymentMethodFieldsOnStep(activeStep);
+    const haveCompletePaymentMethod = paymentMethodFields.some((pmf: {servar: {key: string}}) =>
+      // @ts-expect-error
```
Comment (bo-dun-1) [inline src/form/Form.tsx:969]: can we add a comment to each expect-error so we know why we're leaving an error in?

File: src/form/Form.tsx [hunk 18]
```diff
@@ -953,6 +962,63 @@ function Form({
...
+      Boolean(formattedFields[pmf.servar.key].value?.complete)
```
Comment (bo-dun-1) [inline src/form/Form.tsx:970]: Boolean() isn't necessary here

File: src/form/Form.tsx [hunk 19]
```diff
@@ -953,6 +962,63 @@ function Form({
...
+    );
+    // @ts-expect-error
+    if (integrations.stripe && haveCompletePaymentMethod) {
```
Comment (bo-dun-1) [inline src/form/Form.tsx:973]: all of this logic should live in handleActions below, which was designed to be the entrypoint for integration submissino logic

Comment (bo-dun-1) [reply src/form/Form.tsx:973]: in general we should minimize the amount of stripe-specific logic in Form.tsx, which should be for pure form logic and is already quite hefty.

File: src/form/grid/Cell.jsx [hunk 20]
```diff
@@ -390,6 +394,31 @@ const Cell = ({ node: el, form }) => {
             }}
           />
         );
+      case 'payment_method':
+        return (
+          <Elements.PaymentMethodField
+            {...fieldProps}
+            runtime
+            setCardElement={setCardElement}
+            setFieldError={async (message) => {
```
Comment (bo-dun-1) [inline src/form/grid/Cell.jsx:403]: async/await can be removed, and return setFormElementError from the function directly

File: src/integrations/stripe.js [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+  client,
+  formFieldValues,
+  paymentMethodFieldId,
+  stripeConfig,
+  stripe,
+  cardElement
+) {
+  if (!stripeConfig || Object.keys(stripeConfig).length === 0) return null;
+
+  const { intent_secret: intentSecret } = await client.setupPaymentIntent(
```
Comment (bo-dun-1) [inline src/integrations/stripe.js:82]: can any of this payment method setup logic be moved to the backend? in general whatever we can handle on the backend, we should. it'll also keep the react lib more lightweight

File: webpack.development.js [hunk 22]
```diff
@@ -1,19 +1,19 @@
...
-module.exports = config;
+  config.devtool = 'eval-cheap-module-source-map';
+  config.externals = ['react'];
+  config.plugins.push(
+    new webpack.optimize.LimitChunkCountPlugin({
+      maxChunks: 1
+    })
+  );
+  return config;
+}
```
Comment (bo-dun-1) [inline webpack.development.js:19]: think we need a newline here

File: webpack.development.js [hunk 23]
```diff
@@ -1,19 +1,19 @@
...
-config.plugins.push(
-  new webpack.optimize.LimitChunkCountPlugin({
-    maxChunks: 1
-  })
-);
+module.exports = (env) => {
+  if (env.analyze) {
+    const BundleAnalyzerPlugin =
+      require('webpack-bundle-analyzer').BundleAnalyzerPlugin;
+    config.plugins.push(new BundleAnalyzerPlugin());
```
Comment (bo-dun-1) [inline webpack.development.js:8]: can we include the flags? defaultSizes: 'stat', openAnalyzer: true

**Review by bo-dun-1 — CHANGES_REQUESTED** [review]

File: src/utils/utils.js [hunk 24]
```diff
@@ -0,0 +1,7 @@
+export function filterKeys(obj, allowedKeys) {
```
[the remark this answers is not in the document]
Comment (ericpias) [reply src/utils/utils.js:1]: Moved to primitives

File: src/elements/fields/PaymentMethodField.jsx (hunk 11 above)
Comment (ericpias) [reply src/elements/fields/PaymentMethodField.jsx:89]: It made sense to put in src/elements/styles.

File: src/integrations/stripe.js (hunk 14 above)
Comment (ericpias) [reply src/integrations/stripe.js:15]: Yes, simplification is in order here.  Combining them into one.

File: src/integrations/stripe.js (hunk 16 above)
Comment (ericpias) [reply src/integrations/stripe.js:38]: There is a subtle but important difference between the code and what is suggested.  customer.shipping is not required, but if present, name must be there.  So changed it to be:
if (customer.shipping && !customer.shipping.name) customer.shipping.name = 'Name not supplied'

File: src/integrations/stripe.js (hunk 21 above)
Comment (ericpias) [reply src/integrations/stripe.js:82]: To the best of my knowledge we have moved everything possible to the backend.  The stripe.confirmCardSetup must be done from the client api in order to send the cardElement with all the card info like number, exp. cvc, etc.

File: src/elements/fields/PaymentMethodField.jsx (hunk 9 above)
Comment (ericpias) [reply src/elements/fields/PaymentMethodField.jsx:76]: Changed ApplyStyles.applySelectorStyles just a bit (with a unit test as well) so that the !important is parameterized.  Removed filterStyleValues() completely.

File: src/form/grid/Cell.jsx (hunk 20 above)
Comment (ericpias) [reply src/form/grid/Cell.jsx:403]: Yep, I refactored how I was doing this and left this in.  Fixed.

File: src/elements/fields/PaymentMethodField.jsx (hunk 12 above)
Comment (ericpias) [reply src/elements/fields/PaymentMethodField.jsx:133]: @bo-dun-1 I looked at the input in FileUploadField.  I don't see a lot of commonality other than they both serve as 'anchors' for an html5 error.  The one in FileUploadField actually serves a functional purpose beyond that.  The one in PaymentMethod only serves as the anchor.  So, I just don't see the value in commonizing them.  If you disagree, maybe you can offer some specifics?

File: src/form/Form.tsx (hunk 19 above)
Comment (ericpias) [reply src/form/Form.tsx]: Ok, good to know regarding the purpose of handleActions.  I missed that.  I moved the stripe card setup to there which is cleaner and makes use of the existing design to set the field error, etc.

File: src/elements/fields/PaymentMethodField.jsx (hunk 8 above)
Comment (ericpias) [reply src/elements/fields/PaymentMethodField.jsx:68]: @bo-dun-1 So I looked into this and understand mobile styles and how they work a bit more.  In general, mobile styles are supported for the payment method field.  However, we do this using media queries.  The Stripe card element does not support this so the mobile styles that are set in media queries directly in the card element's style object are (safely) ignored.  This means that mobile styles like color and the various font styles do not work (mobile settings will not be applied).  However, other mobile styles for things like background color and borders are set in the parent of the card element and DO WORK.

I am sure we could make all the mobile styles work using javascript matchMedia, etc. instead of media queries.  However, I am inclined to ignore this issue for now and I have put a comment similar to the above in the payment method component code.  If you think this must instead be addressed/fixed now, let me know.

File: src/form/Form.tsx (hunk 19 above)
Comment (ericpias) [reply src/form/Form.tsx]: Regarding the second comment, I moved the bulk of the code into stripe.js in Integrations in a new method: setupPaymentMethods.

File: src/form/Form.tsx (hunk 17 above)
Comment (ericpias) [reply src/form/Form.tsx]: With other changes I made, much of this code moved out of Form.tsx.  I commented the remaining expected TS error.

File: src/elements/fields/PaymentMethodField.jsx (hunk 8 above)
Comment (bo-dun-1) [reply src/elements/fields/PaymentMethodField.jsx:73]: makes sense! let's ignore for now

File: src/elements/fields/PaymentMethodField.jsx [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+      invalid: {
+        color: ERROR_COLOR,
+        iconColor: ERROR_COLOR
+      }
+    },
+    classes: {
+      base: element.servar.key
+    }
+  };
+  const handleCardChange = async (event) => {
```
Comment (bo-dun-1) [inline src/elements/fields/PaymentMethodField.jsx:100]: this doesn't seem like it needs to be async since no await is being used

File: src/elements/fields/PaymentMethodField.jsx (hunk 12 above)
Comment (bo-dun-1) [reply src/elements/fields/PaymentMethodField.jsx:142]: maybe at least their styles? currently file uses opacity and this one uses background

File: src/form/Form.tsx [hunk 26]
```diff
@@ -1053,7 +1059,31 @@ function Form({
...
+    setCardElementMappings((cardElementMappings) => ({
+      ...cardElementMappings,
+      [key]: cardElement
+    }));
+  }
+
+  async function handleActions(setLoader: any, formattedFields: any) {
+    // Perform any Stripe payment methods setup here
+    try {
+      await setupPaymentMethods(
```
Comment (bo-dun-1) [inline src/form/Form.tsx:1072]: you can't try/catch errors from async functions like this, you'd have to do 

const handleActions = (...) => setupPaymentMethods(...).catch((p) => p);

File: src/integrations/stripe.js [hunk 27]
```diff
@@ -0,0 +1,149 @@
...
+          updateFieldValues({
+            [paymentMethodField.servar.key]: paymentMethodData
+          });
+          formattedFields[paymentMethodField.servar.key] = {
+            value: paymentMethodData,
+            type: paymentMethodField.servar.type,
+            displayText: paymentMethodField.servar.name
+          };
+        } catch (e) {
+          throw Promise.resolve({
```
Comment (bo-dun-1) [inline src/integrations/stripe.js:141]: you should be able to throw the error directly

File: src/form/Form.tsx [hunk 28]
```diff
@@ -166,6 +168,9 @@ function Form({
 
   const [repeats, setRepeats] = useState(0);
 
+  // Stripe - the card elements on the active step, if any
+  const [cardElementMappings, setCardElementMappings] = useState({});
```
Comment (bo-dun-1) [inline src/form/Form.tsx:172]: this state management, the setCardElement function below, and the setupPAymentMethod function initialization imo can all be condensed into a single usePayments custom hook

File: src/integrations/stripe.js [hunk 29]
```diff
@@ -0,0 +1,149 @@
...
+  fieldValues,
+  formattedFields,
+  stripeConfig,
+  cardElementMappings,
+  updateFieldValues,
+  setLoader
+) {
+  // Perform any Stripe payment method setup here if and only if the
+  // payment method field has a complete value
+  const paymentMethodFields = getPaymentMethodFieldsOnStep(activeStep);
```
Comment (bo-dun-1) [inline src/integrations/stripe.js:109]: instead of doing this in the setupPaymentMethods function, handleActions already iterates through all of the step fields to see which ones an action needs to be performed on. i think we should leverage the handleActions logic instead of recreating it here

File: webpack.development.js [hunk 30]
```diff
@@ -1,19 +1,21 @@
...
 const webpack = require('webpack');
 
-config.devtool = 'eval-cheap-module-source-map';
-config.externals = ['react'];
-config.plugins.push(
-  new webpack.optimize.LimitChunkCountPlugin({
-    maxChunks: 1
-  })
-);
+module.exports = (env) => {
```
Comment (bo-dun-1) [inline webpack.development.js:4]: can we pull this bundle analyzer refactor into a separate PR that we can merge first?

File: src/form/Form.tsx (hunk 26 above)
Comment (obmelvin) [reply src/form/Form.tsx:1072]: I might be missing a bit of nuance about what exactly this code is trying to do, but you can use try/catch blocks within an async fn - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#promise_rejection

File: src/integrations/stripe.js (hunk 29 above)
Comment (ericpias) [reply src/integrations/stripe.js:116]: @bo-dun-1 I purposely put this before the login and Pin actions because it seems logical to me that card setup needs to succeed first before and login/pin input attempt.  I think it matters, but maybe not?

File: src/form/Form.tsx (hunk 26 above)
Comment (ericpias) [reply src/form/Form.tsx:1072]: I simplified here to remove the throw and catch.  A simple error return was sufficient.

File: src/integrations/stripe.js (hunk 27 above)
Comment (ericpias) [reply src/integrations/stripe.js:148]: I simplified here to remove the throw and catch. A simple error return was sufficient.

File: src/form/Form.tsx (hunk 28 above)
Comment (ericpias) [reply src/form/Form.tsx:172]: Created the hook function usePayments.

File: src/integrations/stripe.js (hunk 29 above)
Comment (ericpias) [reply src/integrations/stripe.js]: @bo-dun-1 The setup payment method "action" needs to happen for each payment method field on the step and like I said before it seems that it should be done before the login/PIN input actions.  Also, for the login/pin actions, the logic is purposely written to only do one of them if there are multiple.  It just seems these are separate and different purposes here, so I left the setupPaymentMethods() as it was (before the other actions).  If there is a problem setting up a payment method, then the login/pin won't even be attempted and that seems correct to me.

If you really disagree, of course, push back.  @obmelvin and I discussed and he basically agreed.

File: src/elements/fields/PaymentMethodField.jsx [hunk 31]
```diff
@@ -0,0 +1,194 @@
...
+          alignItems: 'center',
+          ...applyStyles.getTarget('sub-fc'),
+          '&:focus': applyStyles.getTarget('active'),
+          '&:hover': applyStyles.getTarget('hover'),
+          ...(inlineError ? { borderColor: ERROR_COLOR } : {})
+        }}
+      >
+        <div css={{ width: '100%', position: 'relative' }}>
+          {/* position an input field under the card element to support html5 errors */}
+          {runtime && errorDisplayMode === 'html5' && (
```
Comment (bo-dun-1) [inline src/elements/fields/PaymentMethodField.jsx:141]: editable is set to true in dashboard and falsey in react lib, so you don't need an additional runtime boolean here

Comment (ericpias) [reply src/elements/fields/PaymentMethodField.jsx:141]: Ah, I missed editable.  Good to know.  Fixed.

File: src/integrations/stripe.js (hunk 29 above)
Comment (ericpias) [reply src/integrations/stripe.js]: As we discussed, I implemented an array of action configurations that groups actions by servar type, orders them, configures continue behavior, etc.  The handleActions() method now has a generic loop to runs through the action configs by servar type and executes what if finds matching fields for.

File: src/elements/fields/PaymentMethodField.jsx [hunk 32]
```diff
@@ -0,0 +1,193 @@
...
+        </div>
+        <InlineTooltip element={element} applyStyles={applyStyles} />
+      </div>
+    </div>
+  );
+};
+
+function PaymentMethodField({ editable = false, ...props }) {
+  useEffect(() => {
+    // if not runtime, i.e. running in dashboard, then just load stripe with key so cardElement shows
```
Comment (bo-dun-1) [inline src/elements/fields/PaymentMethodField.jsx:177]: if editable

File: src/form/Form.tsx [hunk 33]
```diff
@@ -1053,29 +1056,88 @@ function Form({
...
-        setLoader();
-        // Unless we want to do something more complex here we need to make a
-        // choice and can't just return both. Prioritize stytch
-        // @ts-expect-error
-        if (integrations.stytch) {
-          return sendMagicLink(fieldVal);
-        } else {
-          return sendLoginCode(fieldVal, servar);
+  // usePayments /Stripe
+  const [setCardElement, setupPaymentMethod] = usePayments();
```
Comment (bo-dun-1) [inline src/form/Form.tsx:1060]: i think it would be cleaner to pull setupPaymentMethod out of the hook, then you can pull the entire actionConfigurations out into a global const too.

File: src/form/Form.tsx [hunk 34]
```diff
@@ -1053,29 +1056,88 @@ function Form({
...
+        integrationKey: 'firebase',
+        actionFn: ({ fieldVal, servar }: ActionData) =>
+          sendLoginCode(fieldVal, servar),
+        continue: false
+      },
+      {
+        servarType: 'pin_input',
+        integrationKey: 'firebase',
+        isMatch: ({ servar }: ActionData) => servar.metadata.verify_sms_code,
+        actionFn: ({ fieldVal, servar, client }: ActionData) =>
```
Comment (bo-dun-1) [inline src/form/Form.tsx:1097]: you can just modify the helper functions directly to take in ActionData instead of having an intermediate actionFn parse it

File: src/form/Form.tsx (hunk 33 above)
Comment (bo-dun-1) [reply src/form/Form.tsx:1060]: i might put it in integrations/utils

File: src/form/Form.tsx (hunk 34 above)
Comment (ericpias) [reply src/form/Form.tsx:1097]: These functions are called from elsewhere as well.  I was trying to avoid the regression test effort by not changing the signatures, but I guess I will just do it.

File: src/form/Form.tsx (hunk 33 above)
Comment (ericpias) [reply src/form/Form.tsx]: One alternative that I just that thought of is to move targetElementFn out of actionConfigurations and then in handleActions always just set a function called getCardElement into the ActionData (but only stripe would ever use it).  Not perfect, but I think actionConfigurations could then be a static global.  I favor leaving it as is, but if you like this idea  or some other idea better let me know.

**Review by bo-dun-1 — APPROVED** [review]
